mtsp/src/main.leo

```diff
@@ -0,0 +1,792 @@
+  ) -> Future {
+    return finalize_prehook_public(owner, amount, authorized_until, self.caller);
+  }
+
+  async function finalize_prehook_public(
+    owner: TokenOwner,
+    amount: u128,
+    authorized_until: u32,
+    caller: address
+  ) {
+    // Check that the caller has permission to authorize
+    let token: TokenMetadata = registered_tokens.get(owner.token_id);
+    let token_admin: address = token.admin;
+    let external_authorization_party: address = token.external_authorization_party;
+    assert(caller == token_admin || caller == external_authorization_party);
+
+    // Get the balance for the recipient
+    let balance_key: field = BHP256::hash_to_field(owner);
+    let balance: Balance = balances.get(balance_key);
+    // Update the balance, and check that the balance >= 0
+    let new_balance: Balance = Balance {
+      token_id: owner.token_id,
+      account: owner.account,
+      balance: balance.balance - amount,
+      authorized_until: balance.authorized_until
+    };
+    balances.set(balance_key, new_balance);
+
+    // Move balance to authorized_balances
+    let default_balance: Balance = Balance {
+      token_id: owner.token_id,
+      account: owner.account,
+      balance: 0u128,
+      authorized_until: authorized_until
+    };
+    let authorized_balance: Balance = authorized_balances.get_or_use(balance_key, default_balance);
+    let new_authorized_balance: Balance = Balance {
+      token_id: owner.token_id,
+      account: owner.account,
+      balance: authorized_balance.balance + amount,
+      authorized_until: authorized_until
+    };
+    authorized_balances.set(balance_key, new_authorized_balance);
+  }
+
+  async transition prehook_private(input_record: Token, amount: u128, authorized_until: u32) -> (Token, Token, Future) {
+    let unauthorized_record: Token = Token {
+        owner: input_record.owner,
+        amount: input_record.amount - amount,
+        token_id: input_record.token_id,
+        authorized_until: input_record.authorized_until
+    };
+    let authorized_record: Token = Token {
+        owner: input_record.owner,
+        amount: amount,
+        token_id: input_record.token_id,
+        authorized_until: authorized_until
+    };
+    return (unauthorized_record, authorized_record, finalize_prehook_private(input_record.token_id, self.caller));
+  }
+
+  async function finalize_prehook_private(token_id: field, caller: address) {
+    // Check that the caller has permission to authorize
+    let token: TokenMetadata = registered_tokens.get(owner.token_id);
+    let token_admin: address = token.admin;
+    let external_authorization_party: address = token.external_authorization_party;
+    assert(caller == token_admin || caller == external_authorization_party);
+  }
+
+  // -------------------------
+  // Called by token owners/DeFi contracts
+  // -------------------------
+
+  async transition transfer_public(token_id: field, recipient: address, amount: u128) -> Future {
```

Comment:
We should additionally have a `transfer_public_as_signer` that uses `self.signer` instead of `self.caller`